src/cpp/server/server_cc.cc

```diff
@@ -1364,20 +1367,39 @@ grpc::ServerInitializer* Server::initializer() {
   return server_initializer_.get();
 }
 
-grpc::CompletionQueue* Server::CallbackCQ() {
-  // TODO(vjpai): Consider using a single global CQ for the default CQ
-  // if there is no explicit per-server CQ registered
-  grpc::internal::MutexLock l(&mu_);
-  if (callback_cq_ == nullptr) {
+static constexpr unsigned kDefaultCallbackCqShards = 32;
+void GPR_ATTRIBUTE_NOINLINE Server::MakeCallbackCQShardsLocked() {
+  const unsigned num_cores = gpr_cpu_num_cores();
+  const size_t num_cb_cq_shards =
+      num_cores == 0 ? kDefaultCallbackCqShards
+                     : static_cast<size_t>(pow(
+                           2, static_cast<int>(ceil(log(num_cores) / log(2)))));
+  callback_cq_shards_.reserve(num_cb_cq_shards);
+  for (auto idx = 0; idx < num_cb_cq_shards; idx++) {
     auto* shutdown_callback = new grpc::ShutdownCallback;
-    callback_cq_ = new grpc::CompletionQueue(grpc_completion_queue_attributes{
-        GRPC_CQ_CURRENT_VERSION, GRPC_CQ_CALLBACK, GRPC_CQ_DEFAULT_POLLING,
-        shutdown_callback});
+    CompletionQueue* cq =
+        new grpc::CompletionQueue(grpc_completion_queue_attributes{
+            GRPC_CQ_CURRENT_VERSION, GRPC_CQ_CALLBACK, GRPC_CQ_DEFAULT_POLLING,
+            shutdown_callback});
+    callback_cq_shards_.push_back(cq);
 
     // Transfer ownership of the new cq to its own shutdown callback
-    shutdown_callback->TakeCQ(callback_cq_);
+    shutdown_callback->TakeCQ(cq);
   }
-  return callback_cq_;
+}
+
+grpc::CompletionQueue* Server::CallbackCQ() {
+  // TODO(vjpai): Consider using a single global CQ for the default CQ
+  // if there is no explicit per-server CQ registered
```

Comment:
The above TODO is obsoleted by this change, correct?

Reply:
I believe so - a single global would make contention even worse.